src/schedule/dma_single_chan_domain.c

```diff
@@ -0,0 +1,464 @@
+#include <sof/schedule/task.h>
+#include <ipc/topology.h>
+#include <errno.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#define DMA_DOMAIN_OWNER_INVALID	0xFFFFFFFF
+
+struct dma_domain_data {
+	struct notifier notifier;
+	int irq;
+	struct dma_chan_data *channel;
+	void (*handler)(void *arg);
+	void *arg;
+};
+
+struct dma_domain {
+	struct dma *dma_array;	/* pointer to scheduling DMAs */
+	uint32_t num_dma;	/* number of scheduling DMAs */
+	uint32_t owner;		/* core owning the scheduling channel */
+
+	/* data per core */
+	struct dma_domain_data data[PLATFORM_CORE_COUNT];
+};
+
+struct ll_schedule_domain_ops dma_single_chan_domain_ops;
+
+static void dma_single_chan_domain_enable(struct ll_schedule_domain *domain,
+					  int core);
+
+/**
+ * \brief Retrieves DMA channel with lowest period.
+ * \param[in,out] dma_domain Pointer to DMA domain.
+ * \return Pointer to DMA channel.
+ */
+static struct dma_chan_data *dma_chan_min_period(struct dma_domain *dma_domain)
+{
+	struct dma *dmas = dma_domain->dma_array;
+	struct dma_chan_data *channel = NULL;
+	int i;
+	int j;
+
+	/* get currently registered channel if exists */
+	if (dma_domain->owner != DMA_DOMAIN_OWNER_INVALID &&
+	    dma_domain->data[dma_domain->owner].channel)
+		channel = dma_domain->data[dma_domain->owner].channel;
+
+	for (i = 0; i < dma_domain->num_dma; ++i) {
+		for (j = 0; j < dmas[i].plat_data.channels; ++j) {
+			/* channel not set as scheduling source */
+			if (!dma_is_scheduling_source(&dmas[i].chan[j]))
+				continue;
+
+			/* channel not running */
+			if (dmas[i].chan[j].status != COMP_STATE_ACTIVE)
+				continue;
+
+			/* bigger period */
+			if (channel &&
+			    channel->period <= dmas[i].chan[j].period)
+				continue;
+
+			channel = &dmas[i].chan[j];
+		}
+	}
+
+	return channel;
+}
+
+/**
+ * \brief Sends notification event about scheduling DMA channel change.
+ * \param[in,out] channel Pointer to new scheduling DMA channel.
+ */
+static void dma_domain_notify_change(struct dma_chan_data *channel)
+{
+	struct notify_data notify_data;
+
+	notify_data.id = NOTIFIER_ID_DMA_DOMAIN_CHANGE;
+	notify_data.target_core_mask =
+		NOTIFIER_TARGET_CORE_ALL_MASK & ~BIT(cpu_get_id());
+	notify_data.data_size = sizeof(*channel);
+	notify_data.data = channel;
+
+	notifier_event(&notify_data);
+}
+
+/**
+ * \brief Registers and enables selected DMA interrupt.
+ * \param[in,out] channel Pointer to DMA channel.
+ * \param[in,out] data Pointer to DMA domain data.
+ * \param[in,out] handler Pointer to DMA interrupt handler.
+ * \param[in,out] arg Pointer to DMA interrupt handler's argument.
+ * \return Error code.
+ */
+static int dma_single_chan_domain_irq_register(struct dma_chan_data *channel,
+					       struct dma_domain_data *data,
+					       void (*handler)(void *arg),
+					       void *arg)
+{
+	int irq = dma_chan_irq(channel->dma, channel->index);
+	int ret;
+
+	data->irq = interrupt_get_irq(irq, dma_irq_name(channel->dma));
```

Comment:
As I mentioned on the already merged commit, the mere usage of a simple dma_irq_name macro which only considers the DMAC itself might cause problems on the i.MX platform, due to the interrupts for a DMAC being possibly spread among multiple cascaded controllers.

Reply:
@paulstelian97 `dma_chan_irq` takes both dma and channel, so I don't see any reasons why not to extend `dma_irq_name` the same way. Feel free to do it.

Reply:
Guess I'll do the change locally and introduce it together with the EDMA itself. Thanks for the suggestion.